core/base_service.go

```diff
@@ -150,60 +151,122 @@ func (service *BaseService) SetUserAgent(userAgentString string) {
 	userAgent := req.Header.Get(USER_AGENT)
 	if userAgent == "" {
 		req.Header.Add(USER_AGENT, service.UserAgent)
 	}
 
 	// Add authentication to the outbound request.
 	if service.Options.Authenticator == nil {
-		return nil, fmt.Errorf(ERRORMSG_NO_AUTHENTICATOR)
+		err = fmt.Errorf(ERRORMSG_NO_AUTHENTICATOR)
+		return
 	}
-	err := service.Options.Authenticator.Authenticate(req)
+
+	err = service.Options.Authenticator.Authenticate(req)
 	if err != nil {
-		return nil, err
+		return
 	}
 
-	// Perform the request.
-	resp, err := service.Client.Do(req)
+	// Invoke the request.
+	httpResponse, err := service.Client.Do(req)
 	if err != nil {
-		return nil, err
+		return
 	}
 
-	response := new(DetailedResponse)
-	response.Headers = resp.Header
-	response.StatusCode = resp.StatusCode
-	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
-		if resp != nil {
-			response.Result = resp
-			message := getErrorMessage(resp)
-			return response, fmt.Errorf(message)
-		}
+	// Start to populate the DetailedResponse.
+	detailedResponse = &DetailedResponse{
+		StatusCode: httpResponse.StatusCode,
+		Headers:    httpResponse.Header,
 	}
 
-	contentType := resp.Header.Get(CONTENT_TYPE)
-	if contentType != "" {
-		if IsJSONMimeType(contentType) && result != nil {
-			json.NewDecoder(resp.Body).Decode(&result)
-			response.Result = result
-			defer resp.Body.Close()
+	contentType := httpResponse.Header.Get(CONTENT_TYPE)
+
+	// If the operation was unsuccessful, then set up the DetailedResponse
+	// and error objects appropriately.
+	if httpResponse.StatusCode < 200 || httpResponse.StatusCode >= 300 {
+
+		var responseBody []byte
+
+		// First, read the response body into a byte array.
+		if httpResponse.Body != nil {
+			var readErr error
+
+			defer httpResponse.Body.Close()
+			responseBody, readErr = ioutil.ReadAll(httpResponse.Body)
+			if readErr != nil {
+				err = fmt.Errorf("An error occurred while reading the response body: '%s'", readErr.Error())
+				return
+			}
+		}
+
+		// If the responseBody is empty, then just return a generic error based on the status code.
+		if len(responseBody) == 0 {
+			err = fmt.Errorf(http.StatusText(httpResponse.StatusCode))
+			return
+		}
+
+		// For a JSON-based error response body, decode it into a map (generic JSON object).
+		if IsJSONMimeType(contentType) {
+			// Return the error response body as a map, along with an
+			// error object containing our best guess at an error message.
+			responseMap, decodeErr := decodeAsMap(responseBody)
+			if decodeErr == nil {
+				detailedResponse.Result = responseMap
+				err = fmt.Errorf(getErrorMessage(responseMap, detailedResponse.StatusCode))
+				return
+			}
 		}
+
+		// For a non-JSON response or if we tripped while decoding the JSON response,
+		// just return the response body byte array in the RawResult field along with
+		// an error object that contains the generic error message for the status code.
+		detailedResponse.RawResult = responseBody
```

Comment:
I am not sure if we really want to put in all the parsed error in the detailed response result. Why don't we deserialize and put it in the error object? This would be more go-like according to me.

```
detailedResponse, err := service.Request(req, new(Foo))
if err != nil {
  panic(err)
}
```

Reply:
Yes, I thought about using a custom error type, but in the examples that I saw, it looked like it was more difficult for users to deal with those custom error types.
At some level, it makes sense to assign the decoded (JSON) response body (for either a successful or unsuccessful invocation) in the "Response" field.   If there was any sort of problem decoding the response body, then the raw byte array is set on RawResult.
The user can rely on the DetailedResponse.StatusCode field to know whether the operation succeeded or not and then use the Result field accordingly.